File: admin/attachment.go
package admin

import (
	"chat/globals"
	"chat/utils"
	"database/sql"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

type AttachmentFile struct {
	Name           string `json:"name"`
	Size           int64  `json:"size"`
	UpdatedAt      string `json:"updated_at"`
	StorageMode    string `json:"storage_mode"`
	PublicURL      string `json:"public_url"`
	Referenced     bool   `json:"referenced"`
	ReferenceCount int64  `json:"reference_count"`
}

func loadAttachmentReferenceCount(db *sql.DB) map[string]int64 {
	rows, err := globals.QueryDb(db, `SELECT data FROM conversation`)
	if err != nil {
		return map[string]int64{}
	}
	defer func(rows *sql.Rows) {
		_ = rows.Close()
	}(rows)

	references := map[string]int64{}
	for rows.Next() {
		var data string
		if err := rows.Scan(&data); err != nil {
			continue
		}

		for _, name := range utils.ExtractAttachmentNames(data) {
			references[name]++
		}
	}

	return references
}

func ListAttachments(db *sql.DB) ([]AttachmentFile, error) {
	items, err := utils.ListConfiguredStoredAttachments()
	if err != nil {
		return nil, err
	}

	references := loadAttachmentReferenceCount(db)
	result := make([]AttachmentFile, 0, len(items))
	for _, item := range items {
		count := references[item.Name]
		result = append(result, AttachmentFile{
			Name:           item.Name,
			Size:           item.Size,
			UpdatedAt:      item.UpdatedAt.Format("2006-01-02 15:04:05"),
			StorageMode:    strings.ToLower(strings.TrimSpace(item.StorageMode)),
			PublicURL:      item.PublicURL,
			Referenced:     count > 0,
			ReferenceCount: count,
		})
	}

	return result, nil
}

func ListAttachmentAPI(c *gin.Context) {
	data, err := ListAttachments(utils.GetDBFromContext(c))
	if err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": false,
			"error":  err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, data)
}

func DeleteAttachmentAPI(c *gin.Context) {
	name := strings.TrimSpace(c.Query("name"))
	if name == "" {
		c.JSON(http.StatusOK, gin.H{
			"status": false,
			"error":  "attachment name is required",
		})
		return
	}

	if err := utils.DeleteConfiguredStoredAttachment(name); err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": false,
			"error":  err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"status": true,
	})
}

File: app/src/admin/api/attachment.ts
import axios from "axios";
import { CommonResponse } from "@/api/common.ts";
import { getErrorMessage } from "@/utils/base.ts";

export type Attachment = {
  name: string;
  size: number;
  updated_at: string;
  storage_mode: string;
  public_url: string;
  referenced: boolean;
  reference_count: number;
};

export async function listAttachments(): Promise<Attachment[]> {
  try {
    const response = await axios.get("/admin/attachment/list");
    return Array.isArray(response.data) ? (response.data as Attachment[]) : [];
  } catch (e) {
    console.warn(e);
    return [];
  }
}

export async function deleteAttachment(name: string): Promise<CommonResponse> {
  try {
    const response = await axios.post("/admin/attachment/delete", null, {
      params: { name },
    });
    return response.data as CommonResponse;
  } catch (e) {
    console.warn(e);
    return { status: false, error: getErrorMessage(e) };
  }
}
